### PR TITLE
Added Golang time.Time timestamp support #200

### LINKED ISSUE #200: Add support for Golang binary timestamps
Golang serializes timestamps (time.Time structs) in the following binary form [0]

- 1 byte - version (1)
- 8 bytes BE signed int - seconds
- 4 bytes BE signed int - nanoseconds
- 2 bytes BE signed int - zone offset in minutes (where -1 == UTC)

Epoch/zero value is January 1, year 1, 00:00:00.000000000 UTC [1]

[0] Ref: https://cs.opensource.google/go/go/+/refs/tags/go1.17.2:src/time/time.go;drc=1d91551b7326383343c7c143a8ac299d0a685289;l=1169

[1] https://cs.opensource.google/go/go/+/refs/tags/go1.17.2:src/time/time.go;drc=1d91551b7326383343c7c143a8ac299d0a685289;l=333

### PR DESCRIPTION
**Pull request by sydp**
Add a new DateTimeValues and DateTimeEpoch implementation for Golang timestamps that have been serialized in a binary form.

Fixes #200

### CODE REVIEW

**Comment by joachimmetz** [comment]
@sydp is this ready for review?

**Comment by sydp** [comment]
Yes if you have spare cycles thanks, I have to add to the main documentation as well but I can do in another PR?

**Comment by joachimmetz** [comment]
>  I have to add to the main documentation as well but I can do in another PR?

I opt to add the documentation to this PR, you can find it under docs/

**Comment by sydp** [comment]
> > I have to add to the main documentation as well but I can do in another PR?
> 
> I opt to add the documentation to this PR, you can find it under docs/

Sure, will add to it tonight/tomorrow.

**Comment by joachimmetz** [comment]
thx, no hurry, just easier to do all changes at once

File: dfdatetime/golang_time.py [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+      seconds=None, nanoseconds=None, time_zone_offset=None,
+      time_zone_seconds=None):
+    """Initializes a Golang time.Time timestamp.
+
+    Args:
+      seconds (Optional[int]): seconds since epoch.
+      nanoseconds (Optional[int]): nanoseconds component.
+      time_zone_offset (Optional[int]): the time zone in minutes, -1 is a
+          special value for UTC (no Location set).
+      time_zone_seconds (Optional[int]): the time zone in seconds. Only valid
```
Comment (joachimmetz) [inline dfdatetime/golang_time.py:71]: which time zone has an offset in seconds? The rest of of the code also does not support this, are these additional changes that need to be made?

File: dfdatetime/golang_time.py [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+
+  def __init__(self,
+      seconds=None, nanoseconds=None, time_zone_offset=None,
+      time_zone_seconds=None):
+    """Initializes a Golang time.Time timestamp.
+
+    Args:
+      seconds (Optional[int]): seconds since epoch.
+      nanoseconds (Optional[int]): nanoseconds component.
+      time_zone_offset (Optional[int]): the time zone in minutes, -1 is a
```
Comment (joachimmetz) [inline dfdatetime/golang_time.py:69]: So this overrides the interface. Why not pass the 15 or 16 byte value as "timestamp" and parse it inside dfDateTime ?

File: dfdatetime/golang_time.py [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+
+  A Golang time.Time timestamp contans the number of nanoseconds since
+  January 1, 1 UTC. Depending on the version of the timestamp, the time
+  zone is stored in minutes or seconds relative to UTC.
+
+  A serialized version 1 Golang time.Time timestamp is a 15 byte value
+  that consists of:
+
+  * byte 0 - version as an 8-bit integer.
+  * bytes 1-8 - number of seconds since January 1, 1 as a little-endian signed
```
Comment (joachimmetz) [inline dfdatetime/golang_time.py:30]: in the documentation you say these are in big-endian ?

File: tests/golang_time.py [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+
+  def testGetNormalizedTimestamp(self):
+    """Test the _GetNormalizedTimestamp function."""
+    golang_object = golang_time.GolangTime()
+
+    normalized_timestamp = golang_object._GetNormalizedTimestamp()
+    self.assertIsNone(normalized_timestamp)
+
+    golang_object = golang_time.GolangTime(
+        seconds=63772480949, nanoseconds=711098348, time_zone_offset=0
```
Comment (joachimmetz) [inline tests/golang_time.py:55]: No need to but ')' on a separate line, please put these at the end of the previous line

**Review by joachimmetz — CHANGES_REQUESTED** [review]

**Comment by joachimmetz** [comment]
@sydp let's discuss the changes to the time_zone value. IMHO it would be closer to the current interface to pass the 15 or 16 byte value as a single timestamp and have dfDateTime do the parsing/validation of that value.

**Comment by joachimmetz** [comment]
@sydp what is the status of this PR ? Should I take another look ?

File: dfdatetime/golang_time.py (hunk 1 above)
Comment (sydp) [reply dfdatetime/golang_time.py]: As discussed, changed to make v2 raise an error as unsupported.

File: dfdatetime/golang_time.py (hunk 2 above)
Comment (sydp) [reply dfdatetime/golang_time.py]: Changed to parse the bytearray into  time components.

File: tests/golang_time.py (hunk 4 above)
Comment (sydp) [reply tests/golang_time.py]: Fixed.

**Comment by sydp** [comment]
> @sydp what is the status of this PR ? Should I take another look ?

Yes, please have another look.

**Comment by joachimmetz** [comment]
@sydp made some changes, PTAL, but LGTM to merge

**Review by joachimmetz — APPROVED** [review]
LGTM

**Comment by sydp** [comment]
Sorry just clocked off, thanks for the updates, lgtm too 😀

**Pull request merged**